semgrep/semgrep/rule.py

```diff
@@ -48,7 +47,7 @@ def _parse_boolean_expression(
                 if operator in set(OPERATORS_WITH_CHILDREN):
                     if isinstance(pattern_text, list):
                         sub_expression = self._parse_boolean_expression(
-                            pattern_text, 0, f"{prefix}.{pattern_id}"
+                            pattern_text, 0, f"{prefix}.{rule_index}.{pattern_id}"
```

Comment:
This was the bug. Multiple patterns could end up having the same `pattern_id`, which was confusing `semgrep` when it would receive the results back from `semgrep-core`. Since multiple patterns would have the same `pattern_id` then we'd occasionally return invalid results.